### DISCUSSION TITLE
Trouble using SSL to authenticate and secure my Client/Server connection

### DISCUSSION THREAD

**Discussion by dufrtss** [FreeOpcUa/opcua-asyncio#1518] [Q&A]
Hi there,

I'm getting this error: my certificate does not contain the application URI (urn:{host_name}:freeopcua:python:server) even though it is registered as my subject alt name.

Here is the error that is displayed.
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/8ac5e248-bccb-493a-9a03-acecdb546c7d)

Here is the server that I'm using.
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/0115307c-6e0e-4e9c-9e9b-e779e5b401e3)

Here is the subject alt name registered on my certificate.
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/b66fc607-906f-4c1f-8080-03b280fa77c0)

**Comment by dufrtss** [comment]
The client is not recognizing it either even though they both have the subject alt name registered, what am I doing wrong?

![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/8272cfea-7db4-4a7c-a3e1-073930076c52)

**Comment by schroeder-** [comment]
The uri must be the first parameter in the altname. This written in the spec. So some server will enforce it.

**Reply by dufrtss** [reply]
I'm still getting the same errors.

My server upon instantiation and client connection:
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/936c9da4-7098-4b90-913b-debaab83a228)

My client upon connection:
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/2ba551f8-f785-4ef6-9488-42f41f7b618b)

The DNS on the server certificate:
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/97caa0f4-4929-4bef-ac8f-9404fc0ee291)

I couldn't find much on it, sorry but my only cue is discussing this through an issue, I don't know where else to go. I'm trying different approaches, I generated my certificates and private keys through openssl cli commands, my PKI is pretty much just a server and a client certificates and keys, and the server is my CA, which the client is signed to.

**Reply by schroeder-** [reply]
Your cert has DNS:URI:Urn.... instead of URI:urn...

**Reply by dufrtss** [reply]
Thank you, this makes sense, I abstracted my alt name and forgot to update from DNS to URI, instead, I got the two.

**Reply by dufrtss** [reply]
I seem to be getting this error now pointing out that the certificate is not trusted when I try to connect with the client, I signed my client certificate to the server, which is the CA. I don't know what else I could be doing wrong.

![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/b57dff13-dcbb-4572-8d66-bec52b7a5b6c)

**Reply by dufrtss** [reply]
As you can see, here is my client certificate issued by the server.
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/fee46d47-44ed-424b-bbfa-bf10d2132fce)

**Reply by schroeder-** [reply]
Can you add more of your source code? Because from the log, it looks like you are using a certificate validator. Which works as follows

TrustStore is used to validate certificates in two ways:
    - Based on being absent in provided certificate revocation lists
    - The certificate or its issuer being  present in a list of trusted certificates

**Reply by dufrtss** [reply]
Yes, of course, you are being so helpful I appreciate it. Here is the server code:

```
import asyncio
import logging
import socket
import sys
from pathlib import Path

sys.path.insert(0, "..")
from asyncua import Server, ua
from asyncua.crypto.permission_rules import SimpleRoleRuleset
from asyncua.crypto.truststore import TrustStore
from asyncua.crypto.validator import (CertificateValidator,
                                      CertificateValidatorOptions)
from asyncua.server.user_managers import CertificateUserManager

logging.basicConfig(level=logging.INFO)

USE_TRUST_STORE = False

async def main():
    cert_base = Path(__file__).parent
    server_cert = Path(cert_base / "certificates/server/server_cert.der")
    server_private_key = Path(cert_base / "certificates/server/server_private_key.pem")

    host_name = socket.gethostname()
    server_app_uri = f"urn:{host_name}:freeopcua:python:server"
    
    cert_user_manager = CertificateUserManager()
    await cert_user_manager.add_user("certificates/client/client_cert.der", name="test_client")
    
    server = Server(user_manager=cert_user_manager)
    
    await server.init()
    
    await server.set_application_uri(server_app_uri)
    server.set_endpoint("opc.tcp://0.0.0.0:4840/freeopcua/server/")
    server.set_security_policy([ua.SecurityPolicyType.Basic256Sha256_SignAndEncrypt],
                               permission_ruleset=SimpleRoleRuleset())
    
    await server.load_certificate(str(server_cert))
    await server.load_private_key(str(server_private_key))
    
    if USE_TRUST_STORE:
        trust_store = TrustStore([Path('certificates') / 'server' / 'client'], [])
        await trust_store.load()
        
        validator = CertificateValidator(options=CertificateValidatorOptions.TRUSTED_VALIDATION | CertificateValidatorOptions.PEER_CLIENT,
                                         trust_store=trust_store)
    else:
        validator = CertificateValidator(options=CertificateValidatorOptions.EXT_VALIDATION | CertificateValidatorOptions.PEER_CLIENT)
    
    server.set_certificate_validator(validator)
    
    idx = 0

    # Populating the address space
    myobj = await server.nodes.objects.add_object(idx, "MyObject")
    myvar = await myobj.add_variable(idx, "MyVariable", 0.0)
    await myvar.set_writable()  # Set MyVariable to be writable by clients
    
    async with server:
        while True:
            await asyncio.sleep(1)
            current_val = await myvar.get_value()
            count = current_val + 0.1
            await myvar.write_value(count)

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(main())
```

I belive my TrustStore is being used to validate based on providing a list of trusted certificates, correct? In there I passed the paths related to my server and client certificates folders.

Here is my client code:

```
import asyncio
import logging
import socket
import sys
from pathlib import Path

sys.path.insert(0, "..")
from asyncua import Client, ua
from asyncua.crypto.cert_gen import setup_self_signed_certificate
from asyncua.crypto.security_policies import SecurityPolicyBasic256Sha256
from asyncua.crypto.truststore import TrustStore
from asyncua.crypto.validator import (CertificateValidator,
                                      CertificateValidatorOptions)
from cryptography.x509.oid import ExtendedKeyUsageOID

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger(__name__)

USE_TRUST_STORE = True

cert_base = Path(__file__).parent
cert = Path(cert_base / "certificates/client/client_cert.der")
private_key = Path(cert_base / "certificates/client/client_private_key.pem")

async def task(loop):
    host_name = socket.gethostname()
    client_app_uri = f"urn:{host_name}:freeopcua:python:client"
    
    client = Client(url="opc.tcp://127.0.0.1:4840/freeopcua/server/")
    client.application_uri = client_app_uri
    
    await client.set_security(
        SecurityPolicyBasic256Sha256,
        certificate=str(cert),
        private_key=str(private_key),
        server_certificate="certificates/server/server_cert.der",
        mode=ua.MessageSecurityMode.SignAndEncrypt
    )

    if USE_TRUST_STORE:
        trust_store = TrustStore([Path('certificates') / 'server' / 'client'], [])
        await trust_store.load()
        
        validator =CertificateValidator(CertificateValidatorOptions.TRUSTED_VALIDATION | CertificateValidatorOptions.PEER_SERVER,
                                        trust_store)
    else:
        validator =CertificateValidator(CertificateValidatorOptions.EXT_VALIDATION | CertificateValidatorOptions.PEER_SERVER)
    
    client.certificate_validator = validator
    
    try:
        async with client:
            objects = client.nodes.objects
            child = await objects.get_child(['0:MyObject', '0:MyVariable'])
            print(await child.get_value())
            await child.set_value(42)
            print(await child.get_value())
    except ua.UaError as exp:
        _logger.error(exp)

def main():
    loop = asyncio.get_event_loop()
    loop.set_debug(True)
    loop.run_until_complete(task(loop))
    loop.close()

if __name__ == "__main__":
    main()
```

My folder struct:
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/87651515/0112a6a2-b1a8-4f45-af48-45714ebd10ee)

If necessary I can show more info on my certificates .der later.

**Reply by schroeder-** [reply]
change your trust stores to the following, then all certs in both folder are trusted.

`TrustStore([Path('certificates') / 'server', Path('certificates') / 'client'])`

Still this is not a good idead normaly you sturcture your trustlist like this:
certificates
--client
----trustlist
-------trusted_server1.der
-------trusted_server2.der
--server
----trustlist
-------trusted_client1.der
-------trusted_client2.der

**Reply by dufrtss** [reply]
Thank you, I'll separate the keys from the certs in a more structured PKI later. I was mainly figuring out how this would work first.

This helped, but I got an error saying valid PEM but no BEGIN CERTIFICATE/END CERTIFICATE delimiters.

This is odd considering my .pem both client and server start with these delimiters. I'm using them as .der though, as preferred in the docs, but I tried changing them to .pem to see what happened and the same ValueError is printed on the console.

**Reply by dufrtss** [reply]
```
Traceback (most recent call last):
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/client.py", line 68, in <module>
    main()
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/client.py", line 64, in main
    loop.run_until_complete(task(loop))
  File "/home/eduardo.freitas/.asdf/installs/python/3.12.1/lib/python3.12/asyncio/base_events.py", line 684, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/client.py", line 42, in task
    await trust_store.load()
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/venv/lib/python3.12/site-packages/asyncua/crypto/truststore.py", line 55, in load
    await self.load_trust()
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/venv/lib/python3.12/site-packages/asyncua/crypto/truststore.py", line 62, in load_trust
    await self._load_trust_location(location)
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/venv/lib/python3.12/site-packages/asyncua/crypto/truststore.py", line 152, in _load_trust_location
    trusted_cert: crypto.X509 = crypto.X509.from_cryptography(await load_certificate(file_name))
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/venv/lib/python3.12/site-packages/asyncua/crypto/uacrypto.py", line 53, in load_certificate
    return x509.load_pem_x509_certificate(content, default_backend())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eduardo.freitas/dev/projects/opcua-asyncio/venv/lib/python3.12/site-packages/cryptography/x509/base.py", line 583, in load_pem_x509_certificate
    return rust_x509.load_pem_x509_certificate(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: Valid PEM but no BEGIN CERTIFICATE/END CERTIFICATE delimiters. Are you sure this is a certificate?
```

The server code is the same as above, is there a way for me to troubleshoot or debug this besides just opening my certificate file and verifying the delimiters? They seem fine, I'm struggling to identify where the problem could be.

**Reply by schroeder-** [reply]
You could modify the code, so that it logs the certificate before loading.
Also this is more a question for the cryptography library than for asyncua.

**Reply by dufrtss** [reply]
Thank you. I appreciate you helping me out. I'll look into this more.

**Reply by dufrtss** [reply]
Upon analysis, it seems my load_certificate is being called twice, and it gets the key on the second call, is my interpretation of the load_certificate incorrect?

In the first call it seems to assign the certificate to itself properly, but then it is called again passing the private key, and it breaks.

Code is meant to help you visualize the output on the terminal I got from printing my way through this:

Here is the asyncua/server/server.py load_certificate function:
```
    async def load_certificate(self, path_or_content: Union[str, bytes], format: str = None):
        """
        load server certificate from file, either pem or der
        """
        print(path_or_content)
        self.certificate = await uacrypto.load_certificate(path_or_content, format)
        print(self.certificate)
```

Here is the asyncua/crypto/uacrypto.py load_certificate function:
```
async def load_certificate(path_or_content: Union[bytes, str, Path], extension: Optional[str] = None):
    if isinstance(path_or_content, str):
        ext = Path(path_or_content).suffix
    elif isinstance(path_or_content, Path):
        ext = path_or_content.suffix
    else:
        ext = ''
    
    print(ext)
    print(path_or_content)
    content = await get_content(path_or_content)
    if ext == ".pem" or extension == 'pem' or extension == 'PEM':
        return x509.load_pem_x509_certificate(content, default_backend())
    else:
        return x509.load_der_x509_certificate(content, default_backend())
```

FInally, the output I get is this, from which I got to these conclusions:
```
/home/eduardo.freitas/dev/projects/opcua-asyncio/certificates/server/server_cert.der // The path_or_content received by the asyncua server
.der // The suffix of the certificate (so far so good) inside the uacrypto func
/home/eduardo.freitas/dev/projects/opcua-asyncio/certificates/server/server_cert.der // The path_or_content inside the uacrypto func
<Certificate(subject=<Name(C=BR,ST=SC,L=Joinville,O=galax.ia,CN=OPC-UA Server)>, ...)> // Certificate is assigned to self correctly!
.pem // Then it loads again and gets the suffix for my private key?
certificates/server/server_private_key.pem // Path seems to be in the same format it is passed in the moment of creation of the certificate via openssl cmd
```

I looked up any documentations on this but couldn't find anything helpful, which is why I'm recurring to this discussion, if there are any ways I can pursue this on my own, I'd appreciate if you could let me know so I can stop disturbing.

**Reply by dufrtss** [reply]
Still struggling with this.

**Reply by schroeder-** [reply]
It looks like one certificate in the truststore  you specify is some how invalid. Try to clean the truststore, that it only contains public certificates of the clients, that are allowed to connect.

**Reply by dufrtss** [reply]
Hey, this worked out as well, I had to isolate my certs in a trusted folder, found some other issues but it seems to be working now. I'll do some testing tomorrow and if it works out I'll consider it solved and close this discussion. Thank you so much!
